datafusion/core/tests/sqllogictests/src/normalize.rs

```diff
@@ -15,61 +15,86 @@
+                ),
+            )));
+        }
+        rows.append(&mut convert_batch(batch)?);
+    }
+
+    Ok(DBOutput::Rows { types, rows })
+}
 
-/// Normalizes the content of a RecordBatch prior to printing.
+/// Convert a single batch to a `Vec<Vec<String>>` for comparison
+fn convert_batch(batch: RecordBatch) -> Result<Vec<Vec<String>>> {
+    (0..batch.num_rows())
+        .map(|row| {
+            batch
+                .columns()
+                .iter()
+                .map(|col| cell_to_string(col, row))
+                .collect::<Result<Vec<String>>>()
+        })
+        .collect()
+}
+
+/// Normalizes the content of a single cell in RecordBatch prior to printing.
 ///
 /// This is to make the output comparable to the semi-standard .slt format
 ///
 /// Normalizations applied to [NULL Values and empty strings]
 ///
 /// [NULL Values and empty strings]: https://duckdb.org/dev/sqllogictest/result_verification#null-values-and-empty-strings
-pub fn normalize_batch(batch: RecordBatch) -> RecordBatch {
-    let new_columns = batch
-        .columns()
-        .iter()
-        .map(|array| {
-            match array.data_type() {
-                DataType::Utf8 => {
-                    let arr: StringArray = as_string_array(array.as_ref())
-                        .iter()
-                        .map(normalize_string)
-                        .collect();
-                    Arc::new(arr) as ArrayRef
-                }
-                DataType::LargeUtf8 => {
-                    let arr: LargeStringArray = as_largestring_array(array.as_ref())
-                        .iter()
-                        .map(normalize_string)
-                        .collect();
-                    Arc::new(arr) as ArrayRef
-                }
-                // todo normalize dictionary values
+///
+pub fn cell_to_string(col: &ArrayRef, row: usize) -> Result<String> {
+    // represent any null value with the string "NULL"
+    if !col.is_valid(row) {
+        return Ok("NULL".into());
+    }
 
-                // no normalization on this type
-                _ => array.clone(),
-            }
-        })
-        .collect();
+    // Convert to normal string representation
+    let s = arrow::util::display::array_value_to_string(col, row)
+        .map_err(DFSqlLogicTestError::Arrow)?;
 
-    RecordBatch::try_new(batch.schema(), new_columns).expect("creating normalized batch")
-}
+    // apply subsequent normalization depending on type
 
-fn normalize_string(v: Option<&str>) -> Option<&str> {
-    v.map(|v| {
-        // All empty strings are replaced with this value
-        if v.is_empty() {
-            "(empty)"
-        } else {
-            v
+    let s = match col.data_type() {
+        DataType::Utf8 | DataType::LargeUtf8 if s.is_empty() => {
+            // All empty strings are replaced with this value
+            "(empty)".to_string()
         }
-    })
+        // no other normalization
+        _ => s,
+    };
```

Comment:
I think the following is more simple and reduces assignment.
```suggestion
    let mut s = arrow::util::display::array_value_to_string(col, row)
        .map_err(DFSqlLogicTestError::Arrow)?;

    // apply subsequent normalization depending on type

    if matches!(col.data_type(), DataType::Utf8 | DataType::LargeUtf8) && s.is_empty() {
        // All empty strings are replaced with this value
        s = "(empty)".to_string()
    }
```